transports/tcp/src/lib.rs

```diff
@@ -734,116 +714,84 @@ where
-                                        "Failure polling interfaces: {:?}. Scheduling retry.",
-                                        err
-                                    };
-                                    me.pause = Some(Delay::new(me.sleep_on_error));
-                                    return Poll::Ready(Some(Ok(TcpListenerEvent::Error(err))));
-                                }
-                            }
-                        }
-                    }
-                },
-                // If the listener is bound to a single interface, make sure the
-                // address is registered for port reuse and reported once.
-                InAddr::One { addr, out } => {
-                    if let Some(multiaddr) = out.take() {
-                        me.port_reuse.register(*addr, me.listen_addr.port());
-                        return Poll::Ready(Some(Ok(TcpListenerEvent::NewAddress(multiaddr))));
-                    }
+        if let Some(mut pause) = me.pause.take() {
+            match pause.poll_unpin(cx) {
+                Poll::Ready(_) => {}
+                Poll::Pending => {
+                    me.pause = Some(pause);
+                    return Poll::Pending;
                 }
             }
+        }
 
-            if let Some(mut pause) = me.pause.take() {
-                match Pin::new(&mut pause).poll(cx) {
-                    Poll::Ready(_) => {}
-                    Poll::Pending => {
-                        me.pause = Some(pause);
-                        return Poll::Pending;
+        if let Some(if_watcher) = me.if_watcher.as_mut() {
+            loop {
+                match if_watcher.poll_if_event(cx) {
+                    Poll::Ready(Ok(IfEvent::Up(inet))) => {
+                        let ip = inet.addr();
+                        if me.listen_addr.is_ipv4() == ip.is_ipv4() {
+                            let ma = ip_to_multiaddr(ip, me.listen_addr.port());
+                            log::debug!("New listen address: {}", ma);
+                            me.port_reuse.register(ip, me.listen_addr.port());
+                            return Poll::Ready(Some(Ok(TcpListenerEvent::NewAddress(ma))));
+                        }
+                    }
+                    Poll::Ready(Ok(IfEvent::Down(inet))) => {
+                        let ip = inet.addr();
+                        if me.listen_addr.is_ipv4() == ip.is_ipv4() {
+                            let ma = ip_to_multiaddr(ip, me.listen_addr.port());
+                            log::debug!("Expired listen address: {}", ma);
+                            me.port_reuse.unregister(ip, me.listen_addr.port());
+                            return Poll::Ready(Some(Ok(TcpListenerEvent::AddressExpired(ma))));
+                        }
+                    }
+                    Poll::Ready(Err(err)) => {
+                        log::debug! {
+                            "Failure polling interfaces: {:?}. Scheduling retry.",
```

Comment:
I know that this is existing code so we don't need to fix it here but I typically consider logging AND returning an error an anti-pattern.

Reply:
It will lead to printing the same error multiple times.

Reply:
Generally agree. Though in this specific case we only return a very general `TransportError`. The debug output gives some more insight on where the error happened (polling the IfWatcher vs polling the socket). Alternatively we could introduce a new Error enum for the two possible errors.
I am leaning towards just leaving it as it is, but I don't feel strongly about it. Happy to change it if you insist :)

Reply:
It would be nice to not print and return. Once we have proper backtraces, tracing the error back should be no issue.

Until then: Does the return error (`err`) not indicate that it is from the ifwatcher?

Reply:
Good point(s). Removed with 036f56f.